ssl/quic/quic_rx_depack.c

```diff
@@ -0,0 +1,645 @@
+                || (ack_ranges = OPENSSL_zalloc(sizeof(ack_ranges[0])
+                                                * total_ranges)))
+                return 0;
+
+            ack.ack_ranges = ack_ranges;
+            ack.num_ack_ranges = total_ranges;
+
+            if (!ossl_quic_wire_decode_frame_ack(pkt, ack_delay_exp, &ack,
+                                                 NULL))
+                ok = 0;
+            if (ok
+                && mode == process_frames
+                && !ossl_ackm_on_rx_ack_frame(GET_CONN_ACKM(connection), &ack,
+                                              packet_space, received))
+                ok = 0;
+
+            OPENSSL_free(ack_ranges);
+            if (!ok)
+                return 0;
+            break;
+            }
+
+        case OSSL_QUIC_FRAME_TYPE_RESET_STREAM: {
+            OSSL_QUIC_FRAME_RESET_STREAM frame_data;
+
+            if (!ossl_quic_wire_decode_frame_reset_stream(pkt, &frame_data))
+                return 0;
+
+            switch (mode) {
+            case collect_ack_data:
+                /* This frame makes the packet ACK eliciting */
+                ackm_data->is_ack_eliciting = 1;
+                break;
+            case process_frames: {
+                QUIC_STREAM *stream
+                    = ssl_get_stream(connection, frame_data.stream_id);
+                int stream_type = ssl_get_stream_type(stream);
+
+                ssl_close_stream(stream);
+                if ((stream_type & SSL_STREAM_TYPE_W) != 0) {
+                    OSSL_STATM *statm = GET_CONN_STATM(connection);
+
+                    ssl_close_connection(connection);
+                    ossl_statm_set_error_state(statm, QUIC_STREAM_STATE_ERROR);
+
+                    /*
+                     * There's no point to continue with this packet.  However,
+                     * processing it was a success, so we return with a happy
+                     * face, and leave it to the state machine what to do with
+                     * the general STREAM_STATE_ERROR state.
+                     */
+                    return 1;
+                }
+                break;
+                }
+            }
+            break;
+            }
+        case OSSL_QUIC_FRAME_TYPE_STOP_SENDING: {
+            OSSL_QUIC_FRAME_STOP_SENDING frame_data;
+
+            if (!ossl_quic_wire_decode_frame_stop_sending(pkt, &frame_data))
+                return 0;
+
+            switch (mode) {
+            case collect_ack_data:
+                /* This frame makes the packet ACK eliciting */
+                ackm_data->is_ack_eliciting = 1;
+                break;
+            case process_frames: {
+                QUIC_STREAM *stream
+                    = ssl_get_stream(connection, frame_data.stream_id);
+                int stream_type = ssl_get_stream_type(stream);
+
+                ssl_close_stream(stream);
+                if ((stream_type & SSL_STREAM_TYPE_R) != 0) {
```

Comment:
The NULL check you suggested a few lines up is the answer, 'cause I imagine that a stream that has received the stop sending frame can't be accessed any more, i.e. `ssl_get_stream()` should return NULL.

Reply:
Not necessarily. The stream probably still exists (perhaps it still has readable data in its buffers - or is still allowed to receive new data event though it is not allowed to send any more). Actually I'm unclear whether it *is* an error to receive a second stop sending frame (e.g. because of a retransmit)...maybe that's allowed and is just ignored??

Reply:
It can definitely be received multiple times if the ack for the previous one is lost. It is a connection error if it is received for non-existent or receiving-only stream.

Reply:
So this error condition should be probably `(stream_type & SSL_STREAM_TYPE_W) == 0`

Reply:
> So this error condition should be probably `(stream_type & SSL_STREAM_TYPE_W) == 0`

Er what???

Reply:
Code has moved around, it's better to restart this discussion on the refactored code